### PR TITLE
Market Insight update

### LINKED ISSUE #93: Update Market Insight logic base on customer feedback
1. add result sort
2. make number of cards authorable
3. Recreate mock data and move mock method to mockapi.js
4. Use api `sortdescrition` as description in the card

### PR DESCRIPTION
**Pull request by JiangLong2019**
Please always provide the [GitHub issue(s)](../issues) your PR is for, as well as test URLs where your change can be observed (before and after):

Fix #93 

Test URLs:
- Before: https://main--icicidirect--aemsites.hlx.live/draft/jiang/market-insight
- After: https://update-market-insight--icicidirect--aemsites.hlx.live/draft/jiang/market-insight

### CODE REVIEW

**Comment by sdmcraft** [comment]
Looks like the PR description is missing the block library link. Could you please add?

File: blocks/market-insight/market-insight.js [hunk 1]
```diff
@@ -84,35 +56,53 @@ function addSocialButtonEvent(button, block) {
   });
 }
 
-async function decorateCards(block, placeholders, previousNode) {
-  const results = await fetchMarketInsightMockData();
+async function decorateCards(block, placeholders, cardCount, previousNode) {
+  const queryObj = await fetchMarketInsightMockData();
```
Comment (sdmcraft) [inline blocks/market-insight/market-insight.js:60]: Can we avoid await? Please see [this comment](https://github.com/aemsites/icicidirect/pull/82/files#r1548731220)

File: blocks/market-insight/market-insight.js [hunk 2]
```diff
@@ -84,35 +56,53 @@ function addSocialButtonEvent(button, block) {
...
+  const results = queryObj.map((el) => {
+    const elArr = el.PublishedOnDate.split(' ');
```
Comment (sdmcraft) [inline blocks/market-insight/market-insight.js:62]: For null safety, can `el.PublishedOnDate` be undefined?

File: blocks/market-insight/market-insight.js [hunk 3]
```diff
@@ -84,35 +56,53 @@ function addSocialButtonEvent(button, block) {
...
-    description.innerHTML = decodeURIComponent(result.description);
+    description.innerHTML = decodeURIComponent(result.ShortDescription ?? '');
     // Cards powerby
     const powerbyDiv = createElement('div', '');
     const powerbyContent = createElement('p', '');
     powerbyContent.textContent = powerBy;
     const publishedOnContent = createElement('p', '');
-    const publishedon = result.publishedon.replaceAll(' ', '-');
-    publishedOnContent.textContent = `${publishedOn} ${publishedon}`;
+    const publishOndate = result.PublishedOn ? result.PublishedOn.replaceAll(' ', '-') : '';
```
Comment (sdmcraft) [inline blocks/market-insight/market-insight.js:104]: I am thinking if we should use [some standard date formatting](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/DateTimeFormat) function here.

**Comment by JiangLong2019** [comment]
@sdmcraft It's a refactoring PR, block library is already there and nothing needs to change.
>Looks like the PR description is missing the block library link. Could you please add?

File: blocks/market-insight/market-insight.js (hunk 3 above)
Comment (JiangLong2019) [reply blocks/market-insight/market-insight.js:104]: I thought the same, but the original site shows like this. I kept the same look&feel.

image

File: blocks/market-insight/market-insight.js (hunk 2 above)
Comment (JiangLong2019) [reply blocks/market-insight/market-insight.js:62]: Good point, let me add the adjustment.

File: blocks/market-insight/market-insight.js (hunk 1 above)
Comment (JiangLong2019) [reply blocks/market-insight/market-insight.js:60]: We have implemented the `observe()` method, after the page loading, no cards are rendered until the end user scrolls to this block. I think what I need to do is add a loading icon or layer by default and remove it when the cards are loaded. What do you think?

**Review by sdmcraft — APPROVED** [review]

**Pull request merged**